Mathlib/RingTheory/Polynomial/Basic.lean

```diff
@@ -0,0 +1,1292 @@
+        simp only [Ne.def, Prod.mk.inj_iff, not_and_or] at hij
+        obtain hi | hj : i < m ∨ j < n :=
+          by
+          rw [or_iff_not_imp_left, not_lt, le_iff_lt_or_eq]
+          rintro (hmi | rfl)
+          · rw [← not_le]
+            intro hnj
+            exact (add_lt_add_of_lt_of_le hmi hnj).ne hij.2.symm
+          · simp only [eq_self_iff_true, not_true, false_or_iff, add_right_inj,
+              not_and_self_iff] at hij
+        · rw [mul_comm]
+          apply P.mul_mem_left
+          exact Classical.not_not.1 (Nat.find_min hf hi)
+        · apply P.mul_mem_left
+          exact Classical.not_not.1 (Nat.find_min hg hj)
+set_option linter.uppercaseLean3 false in
+#align ideal.is_prime_map_C_iff_is_prime Ideal.isPrime_map_C_iff_isPrime
+
+/-- If `P` is a prime ideal of `R`, then `P.R[x]` is a prime ideal of `R[x]`. -/
+theorem isPrime_map_C_of_isPrime {P : Ideal R} (H : IsPrime P) :
+    IsPrime (map (C : R →+* R[X]) P : Ideal R[X]) :=
+  (isPrime_map_C_iff_isPrime P).mpr H
+set_option linter.uppercaseLean3 false in
+#align ideal.is_prime_map_C_of_is_prime Ideal.isPrime_map_C_of_isPrime
+
+theorem is_fg_degreeLe [IsNoetherianRing R] (I : Ideal R[X]) (n : ℕ) :
+    Submodule.Fg (I.degreeLe n) :=
+  isNoetherian_submodule_left.1
+    (isNoetherian_of_fg_of_noetherian _ ⟨_, degreeLe_eq_span_X_pow.symm⟩) _
+#align ideal.is_fg_degree_le Ideal.is_fg_degreeLe
+
+end CommRing
+
+end Ideal
+
+variable {σ : Type v} {M : Type w}
+
+variable [CommRing R] [CommRing S] [AddCommGroup M] [Module R M]
+
+section Prime
+
+variable (σ) {r : R}
+
+namespace Polynomial
+
+-- Porting note: this ordering of the argument dramatically speeds up lean
+theorem prime_C_iff : Prime (C r) ↔ Prime r :=
+  ⟨comap_prime C (evalRingHom (0 : R)) fun r => eval_C, by
+    intro hr
+    have := hr.1
+    rw [← Ideal.span_singleton_prime] at hr⊢
+    · rw [← Set.image_singleton, ← Ideal.map_span]
+      apply Ideal.isPrime_map_C_of_isPrime hr
+    · intro h; apply (this (C_eq_zero.mp h))
+    · assumption⟩
+set_option linter.uppercaseLean3 false in
+#align polynomial.prime_C_iff Polynomial.prime_C_iff
+
+end Polynomial
+
+namespace MvPolynomial
+
+/- Porting note: had to move the heavy inference outside the convert call to stop timeouts.
+Also, many @'s. etaExperiment caused more time outs-/
+private theorem prime_C_iff_of_fintype {R : Type u} (σ : Type v) {r : R} [CommRing R] [Fintype σ] :
+    Prime (C r : MvPolynomial σ R) ↔ Prime r := by
+  let f (d : ℕ) := (finSuccEquiv R d).symm.toMulEquiv
+  let _coe' (d : ℕ) : CoeFun ((MvPolynomial (Fin d) R)[X] ≃* MvPolynomial (Fin (d + 1)) R)
+    (fun _ => (MvPolynomial (Fin d) R)[X] → MvPolynomial (Fin (d + 1)) R) := inferInstance
+  have that (d : ℕ) : @C R (Fin (d+1)) _ r = (f d) (Polynomial.C (@C R (Fin d) _ r)) := by
+    rw [← finSuccEquiv_comp_C_eq_C]; rfl
+  rw [(renameEquiv R (Fintype.equivFin σ)).toMulEquiv.prime_iff]
+  convert_to Prime (C r) ↔ _
```

Comment:
It doesn't help the speed, but 
```suggestion
  convert_to Prime (C r) ↔ _ using 2
```
saves you the `congr!` in the first case. By default `convert_to` uses `using 1`.

Reply:
Done